### PR TITLE
Invert the dependency between hls-plugin-api and ghcide

### PR DESCRIPTION
**Pull request by pepeiborra**
This PR includes changes both to ghcide and HLS to implement the reorg described in https://github.com/haskell/ghcide/issues/936#issuecomment-751437853

To summarise:

- `hls-plugin-api` no longer depends on ghcide.
- `ghcide` now depends on `hls-plugin-api` and exposes:
  - The ghcide HLS plugin
  - The `asGhcIdePlugin` adaptor

The goals are:
- to be able to break the `ghcide` HLS plugin down 
- to rewrite exe:ghcide on top of the HLS plugin model.

The ghcide side is reviewed in https://github.com/haskell/ghcide/pull/963

If this change is accepted there are two further considerations:
- This would be a good moment to merge the 2 repos, so that there is no history loss. 
- `hls-plugin-api` will need to be released to Hackage prior to merging https://github.com/haskell/ghcide/pull/963

### CODE REVIEW

**Review by ndmitchell — APPROVED** [review]

**Comment by jneira** [comment]
sorry for my ignorance, but how can merge both projects? a project administrator is required? would that keep the history of both?

I would do it right now, as proposed by pepe

**Review by jneira — APPROVED** [review]
thanks for the big refactoring

**Comment by pepeiborra** [comment]
> sorry for my ignorance, but how can merge both projects? a project administrator is required? would that keep the history of both?
> 
> I would do it right now, as proposed by pepe

Merging the projects is two things: 

1. Merging the git repositories.
2. Merging the GitHub projects. Since Github does not support project merging as such, it has to be done manually. I think this would entail: 
  a. Transferring all the issues
  b. Sorting out CI
  c. Archiving the ghcide project

**Comment by jneira** [comment]
I will try to help with issues

**Comment by pepeiborra** [comment]
> I will try to help with issues

It shouldn't be too hard, the `hub` cli tool can be used to automate: 

https://hub.github.com/hub-issue.1.html

**Comment by pepeiborra** [comment]
converting to draft to avoid an accidental merge until #702 has landed

**Comment by Ailrun** [comment]
FYI, I think it can be also done with (this official tool)[https://github.com/cli/cli].

**Comment by pepeiborra** [comment]
Reworked to include the ghcide changes and made exe:ghcide run HLS plugins

File: ghcide/exe/Main.hs
Comment (pepeiborra) [inline ghcide/exe/Main.hs:89]: /cc @gdevanla you can add HLS plugin descriptors to this list and they will be loaded in ghcide

**Comment by pepeiborra** [comment]
``` 
Testing / test (8.6.4, windows-latest, false) (pull_request) Failing after 300m 
```
This target seems more unreliable than the rest. @jneira have you noticed this?

**Comment by pepeiborra** [comment]
It did it again: 

```
Testing / test (8.6.4, windows-latest, false) (pull_request) Failing after 295m — test (8.6.4, windows-latest, false)
```

Do you have any insights or shall I just disable it?

**Comment by jneira** [comment]
@pepeiborra let me take a quick look to the error

**Comment by pepeiborra** [comment]
There is no error, it just times out after 5 hours

**Comment by jneira** [comment]
[In this one](https://github.com/haskell/haskell-language-server/runs/1623957691?check_suite_focus=true) we can see 40 failing tests, all about the eval plugin, due to session timed outs.

In the third execution we can see this output from the cradle:

```
 Output from setting up the cradle Cradle {cradleRootDir = "D:\\a\\haskell-language-server\\haskell-language-server\\test\\testdata\\eval", cradleOptsProg = CradleAction: Cabal}


> Build profile: -w ghc-8.6.4 -O1


> In order, the following will be built (use -v for more details):


>  - QuickCheck-2.14.2 (lib) (requires build)


>  - test-0.1.0.0 (lib) (first run)


> Starting     QuickCheck-2.14.2 (lib)


> Building     QuickCheck-2.14.2 (lib)
```

Maybe it has to build quickcheck for some reason and that makes the session time out?

**Comment by jneira** [comment]
This would be the run history: https://github.com/haskell/haskell-language-server/actions?query=workflow%3ATesting+branch%3Ainvert
The last succesful build for ghc-8.6.4 was https://github.com/haskell/haskell-language-server/actions/runs/450520947
I dont see the ghc-8.6.4 specific error before it, and neither the commit of the first failing one: https://github.com/haskell/haskell-language-server/actions/runs/451027804

**Comment by jneira** [comment]
In any case, if the last one https://github.com/haskell/haskell-language-server/runs/1626173853?check_suite_focus=true fails, i think we should disable it, to unblock this

**Comment by pepeiborra** [comment]
The Windows 8.6.4 job is randomly very slow and times out after 6 hrs. This happens 3? out of 4 times, I'm not sure. In any case, it's the only job that displays this behaviour, and it seems rather unreasonable so I am disabling it

**Pull request merged**
